proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+}
+```
+
+In this scenario, the following classes would be equivalent:
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class ValidatedRepository<A>(with val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        return storedUsers.filter { it.value.isValid() }
+    }
+
+    override fun loadById(id: Int): A? {
+        return storedUsers[id]?.let { if (it.isValid()) it else null }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+
+// Regular Kotlin
+class ValidatedRepository<A>(val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
+
+    override fun loadAll(): Map<Int, A> {
+        with (V) {
+            return storedUsers.filter { it.value.isValid() }
+        }
+    }
+
+    override fun loadById(id: Int): A? {
+        with (V) {
+            return storedUsers[id]?.let { if (it.isValid()) it else null }
+        }
+    }
+
+    override fun A.save() {
+        storedUsers[generateKey(this)] = this
+    }
+}
+```
+
+As you can see in the first example, `A.isValid()` becomes available inside the method’s scope automatically. The equivalent version of doing this without KEEP-87 is to manually add `with (V)` inside each method, as you can see in the second example.  
+
+## Composition and chain of evidences
+
+Constraint interface declarations and extension evidences can encode further constraints on their type parameters so that they can be composed nicely:
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+import com.domain.Group
+
+extension class GroupRepository<A>(with val repoA: Repository<A>) : Repository<Group<A>> {
+  override fun loadAll(): Map<Int, Group<A>> {
+    return repoA.loadAll().mapValues { Group(it.value) }
+  }
+
+  override fun loadById(id: Int): Group<A>? {
+    return repoA.loadById(id)?.let { Group(it) }
+  }
+
+  override fun Group<A>.save() {
+    this.items.map { repoA.run { it.save() } }
+  }
+}
+```
+
+The above extension provides evidence of a `Repository<Group<A>>` as long as there is a `Repository<A>` in scope. The Call site would look like:
+
+```kotlin
+fun <A> fetchGroup(with repo: GroupRepository<A>) = loadAll()
+
+fun main() {
+  fetchGroup<User>() // Succeeds! There's evidence of Repository<Group<A>> and Repository<User> provided in scope.
+  fetchGroup<Coin>() // Fails! There's evidence of Repository<Group<A>> but no evidence of `Repository<Coin>` available.
```

Comment:
`There's evidence of Repository<Group<A>>`

Is this actually checked? 

`with repo: GroupRepository<A>` is injecting an implementation of `Repository<Group<A>>` already. I'd assume that only `Repository<User>` is checked.

If we were doing `fun <A> fetchGroup(with repo: Repository<Group<A>>) = loadAll()` the comment would make sense.

Not sure if i understood it correctly, maybe I am missing something. And sorry for my DI language 😅